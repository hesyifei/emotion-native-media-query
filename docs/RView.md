# `RView` API Reference

This document lays out the current public properties and methods for `RView`.

## Props Index

- [`style`](#style)
- [`rStyle`](#rstyle)
  - [`[MediaRule.MinWidth]`](#mediaruleminwidth)
  - [`[MediaRule.MaxWidth]`](#mediarulemaxwidth)
- [`WebTag`](#webtag)
- [`NativeTag`](#nativetag)

---

# Reference

## Props

### `style`

The (default) style of the view. Only [object styles](https://emotion.sh/docs/object-styles) are supported (i.e., ` css`` ` with string is not supported).

**Example:**

```tsx
import RView, { MediaRule } from "emotion-native-media-query";

<RView
  style={{
    margin: 15,
  }}
/>;
```

---

### `rStyle`

The responsive style of the view. It should contain one or more of the following object:

#### `[MediaRule.MinWidth]`

It is analogous to CSS's `@media (min-width: ___px)`.

Note that, unlike media queries, the order of the widths here doesn't matter as we internally sort the width ascendingly to make sure the correct style is selected.

**Example:**

```tsx
import RView, { MediaRule } from "emotion-native-media-query";

<RView
  style={{
    padding: 11,
  }}
  rStyle={{
    [MediaRule.MinWidth]: {
      // @media (min-width: 200px)
      200: {
        padding: 33,
      },
      // @media (min-width: 500px)
      500: {
        padding: 66,
      },
      // @media (min-width: 800px)
      800: {
        padding: 99,
      },
    },

    // Other media rules
    [MediaRule.MaxWidth]: {
      // ...
    },
  }}
/>;
```

#### `[MediaRule.MaxWidth]`

It is analogous to CSS's `@media (max-width: ___px)`.

Note that, unlike media queries, the order of the widths here doesn't matter as we internally sort the width descendingly to make sure the correct style is selected.

**Example:**

```tsx
import RView, { MediaRule } from "emotion-native-media-query";

<RView
  style={{
    backgroundColor: "black",
  }}
  rStyle={{
    [MediaRule.MaxWidth]: {
      // @media (max-width: 300px)
      300: {
        backgroundColor: "red",
      },
      // @media (max-width: 600px)
      600: {
        backgroundColor: "green",
      },
      // @media (max-width: 900px)
      900: {
        backgroundColor: "blue",
      },
    },

    // Other media rules
    [MediaRule.MinWidth]: {
      // ...
    },
  }}
/>;
```

---

### `WebTag`

The tag used on web. Should be a `React.ElementType` that supports `className` property (for example, `"header"` or `"section"`).

(While it currently also supports component like `Text`, `className` has been deprecated in those components and you should avoid using it.)

**Example:**

```tsx
import RView, { MediaRule } from "emotion-native-media-query";

<RView WebTag="footer" style={/* ... */} rStyle={/* ... */} />;
```

---

### `NativeTag`

The tag used in the native app. Should be a `React.ElementType`.

**Example:**

```tsx
import RView, { MediaRule } from "emotion-native-media-query";

<RView NativeTag={Text} style={/* ... */} rStyle={/* ... */} />;
```
