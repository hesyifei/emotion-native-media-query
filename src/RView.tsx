/** @jsx jsx */
/**
 * @class RView
 */

import * as React from "react";
import { View, Platform, ViewProps } from "react-native";
import { jsx } from "@emotion/core";
import {
  MediaRule,
  RStyle,
  RMinWidthStyle,
  RMaxWidthStyle,
  Style
} from "./types";
import {
  isWidthGreaterThanOrEqualTo,
  isWidthSmallerThanOrEqualTo,
} from "./helpers";

function _minWidthStyleForEach(
  minWidthStyle: RMinWidthStyle,
  callback: (minWidth: number, style: Style) => void
) {
  Object.keys(minWidthStyle)
    // We have to sort the style ascendingly first,
    // else on native, different order will produce different results.
    .sort((a, b) => Number(a) - Number(b))
    .forEach(minWidth => {
      let value = minWidthStyle[minWidth];
      callback(Number(minWidth), value);
    });
}

function _maxWidthStyleForEach(
  maxWidthStyle: RMaxWidthStyle,
  callback: (maxWidth: number, style: Style) => void
) {
  Object.keys(maxWidthStyle)
    // We have to sort the style descendingly first,
    // else on web,
    // - object in JavaScript with number as key will always get sorted ascendingly
    // - and the latest max-width will always dominate any other rule.
    // and on native,
    // - different order will produce different results.
    .sort((a, b) => Number(b) - Number(a))
    .forEach(maxWidth => {
      let value = maxWidthStyle[maxWidth];
      callback(Number(maxWidth), value);
    });
}

function _getStyleWithMediaQuery(
  rStyle: RStyle
): { [mediaQuery: string]: any } {
  const {
    [MediaRule.MinWidth]: minWidthStyle = {},
    [MediaRule.MaxWidth]: maxWidthStyle = {}
  } = rStyle;

  let rStyleMediaQueries = {};
  _minWidthStyleForEach(minWidthStyle, (minWidth, value) => {
    rStyleMediaQueries[`@media (min-width: ${minWidth}px)`] = value;
  });
  _maxWidthStyleForEach(maxWidthStyle, (maxWidth, value) => {
    rStyleMediaQueries[`@media (max-width: ${maxWidth}px)`] = value;
  });

  return rStyleMediaQueries;
}

function _getFlattenedStyleForCurrentScreen(rStyle: RStyle): Style {
  const {
    [MediaRule.MinWidth]: minWidthStyle = {},
    [MediaRule.MaxWidth]: maxWidthStyle = {}
  } = rStyle;

  let style = {};

  _minWidthStyleForEach(minWidthStyle, (minWidth, value) => {
    if (isWidthGreaterThanOrEqualTo(minWidth)) {
      style = { ...style, ...value };
    }
  });

  _maxWidthStyleForEach(maxWidthStyle, (maxWidth, value) => {
    if (isWidthSmallerThanOrEqualTo(maxWidth)) {
      style = { ...style, ...value };
    }
  });

  return style;
}

interface RViewProps extends ViewProps {
  WebTag?: React.ElementType;
  NativeTag?: React.ElementType;
  style?: Style;
  rStyle?: RStyle;
  // Allow arbitrary props too.
  [key: string]: any;
}

const RView: React.FunctionComponent<RViewProps> = ({
  WebTag = "div",
  NativeTag = View,
  style: defaultStyle = {},
  rStyle = {},
  ...remainingProps
}) => {
  if (Platform.OS === "web") {
    // Partly based on https://github.com/necolas/react-native-web/blob/e810f1fd2b41293cb1efe04e332fb6f8d4bcca65/packages/react-native-web/src/exports/View/index.js#L80-L94
    const reactNativeWebViewStyle: Style = {
      display: "flex",
      flexBasis: "auto",
      flexDirection: "column",
      flexShrink: 0,
      margin: 0,
      minHeight: 0,
      minWidth: 0,
      padding: 0,
      position: "relative",
      zIndex: 0
    };

    const rStyleMediaQueries = _getStyleWithMediaQuery(rStyle);
    return (
      <WebTag
        css={{
          ...reactNativeWebViewStyle,
          ...defaultStyle,
          ...rStyleMediaQueries
        }}
        {...remainingProps}
      />
    );
  } else {
    const responsiveStyle = _getFlattenedStyleForCurrentScreen(rStyle);
    return (
      <NativeTag
        style={{
          ...defaultStyle,
          ...responsiveStyle
        }}
        {...remainingProps}
      />
    );
  }
};
export default RView;
