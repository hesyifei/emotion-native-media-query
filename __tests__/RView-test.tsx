import * as React from "react";
import { Text } from "react-native";
import renderer from "react-test-renderer";
import RView from "../src";

describe("RView", () => {
  it("simple", () => {
    const component = renderer.create(
      <RView>
        <Text>Hello world!</Text>
      </RView>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
