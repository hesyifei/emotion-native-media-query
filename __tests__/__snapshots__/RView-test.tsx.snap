// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RView simple 1`] = `
<View
  style={Object {}}
>
  <Text>
    Hello world!
  </Text>
</View>
`;
